internal/semantic/version-redhat.go

```diff
@@ -0,0 +1,235 @@
+		for {
+			if vi == len(a) || !shouldBeTrimmed(rune(a[vi])) {
+				break
+			}
+			vi++
+		}
+
+		for {
+			if wi == len(b) || !shouldBeTrimmed(rune(b[wi])) {
+				break
+			}
+			wi++
+		}
+
+		// 2. If both strings start with a tilde, discard it and move on to the next character.
+		vStartsWithTilde := vi < len(a) && a[vi] == '~'
+		wStartsWithTilde := wi < len(b) && b[wi] == '~'
+
+		if vStartsWithTilde && wStartsWithTilde {
+			vi++
+			wi++
+
+			continue
+		}
+
+		// 3. If string `a` starts with a tilde and string `b` does not, return -1 (string `a` is older); and the inverse if string `b` starts with a tilde and string `a` does not.
+		if vStartsWithTilde {
+			return -1
+		}
+		if wStartsWithTilde {
+			return +1
+		}
+
+		// 4. If both strings start with a caret, discard it and move on to the next character.
+		vStartsWithCaret := vi < len(a) && a[vi] == '^'
+		wStartsWithCaret := wi < len(b) && b[wi] == '^'
+
+		if vStartsWithCaret && wStartsWithCaret {
+			vi++
+			wi++
+
+			continue
+		}
+
+		// 5. if string `a` starts with a caret and string `b` does not, return -1 (string `a` is older) unless string `b` has reached zero length, in which case return +1 (string `a` is newer); and the inverse if string `b` starts with a caret and string `a` does not.
+		if vStartsWithCaret {
+			if wi == len(b) {
+				return +1
+			}
+
+			return -1
+		}
+		if wStartsWithCaret {
+			if vi == len(a) {
+				return -1
+			}
+
+			return +1
+		}
+
+		// 6. End the loop if either string has reached zero length.
+		if vi == len(a) || wi == len(b) {
+			break
+		}
+
+		// 7. If the first character of `a` is a digit, pop the leading chunk of continuous digits from each string (which may be "" for `b` if only one `a` starts with digits). If `a` begins with a letter, do the same for leading letters.
+		isDigit := isASCIIDigit(rune(a[vi]))
+
+		var iser func(r rune) bool
+		if isDigit {
+			iser = isASCIIDigit
+		} else {
+			iser = isASCIILetter
+		}
+
+		var ac, bc string
+
+		for _, c := range a[vi:] {
+			if !iser(c) {
+				break
+			}
+
+			ac += string(c)
+			vi++
+		}
+
+		for _, c := range b[wi:] {
+			if !iser(c) {
+				break
+			}
+
+			bc += string(c)
+			wi++
+		}
+
+		// 8. If the segment from `b` had 0 length, return 1 if the segment from `a` was numeric, or -1 if it was alphabetic. The logical result of this is that if `a` begins with numbers and `b` does not, `a` is newer (return 1). If `a` begins with letters and `b` does not, then `a` is older (return -1). If the leading character(s) from `a` and `b` were both numbers or both letters, continue on.
+		if bc == "" {
+			if isDigit {
+				return +1
+			}
+
+			return -1
+		}
+
+		// 9. If the leading segments were both numeric, discard any leading zeros and whichever one is longer wins. If `a` is longer than `b` (without leading zeroes), return 1, and vice versa. If they’re of the same length, continue on.
+		if isDigit {
```

Comment:
Hmm... would this be clearer to just parse and compare numerically?

Reply:
I don't think so - I tried it previously when doing some performance testing, and it actually broke the tests so I think at best it'll require the same amount of code 🤷‍♂️